File: algo/job-exam/base-sort-algo/quick_sort.py
import random

"""
快排
"""
def quick_sort(arr: list) -> None:
    def sort(low, high):
        if low >= high:
            return
        # 随机选择枢轴
        tmp = random.randint(low, high)
        arr[low], arr[tmp] = arr[tmp], arr[low]

        pivot, i, j = arr[low], low, high
        while i < j:
            while i < j and arr[j] >= pivot:
                j -= 1
            arr[i] = arr[j]
            while i < j and arr[i] <= pivot:
                i += 1
            arr[j] = arr[i]
        arr[i] = pivot
        sort(low, i - 1)
        sort(i + 1, high)

    sort(0, len(arr) - 1)

if __name__ == '__main__':
    num = [5, 4, 3, 2, 1]
    quick_sort(num)

File: job/src/sort/QuickSort.java
package sort;

import java.util.Random;

/**
 * @author dongzhenxun
 * @date 2025/4/23 下午7:42
 * @description 快排
 */
public class QuickSort {

    public static void quickSort(int[] arr) {
        sort(arr, 0, arr.length - 1);
    }

    private static void sort(int[] arr, int low, int high) {
        if (low >= high) {
            return;
        }

        Random random = new Random();
        int pivotIndex = low + random.nextInt(high - low + 1);
        int tmp = arr[pivotIndex];
        arr[pivotIndex] = arr[low];
        arr[low] = tmp;

        int pivot = arr[low];
        int i = low, j = high;
        while (i < j) {
            while (i < j && arr[j] >= pivot) {
                j--;
            }
            arr[i] = arr[j];
            while (i < j && arr[i] <= pivot) {
                i++;
            }
            arr[j] = arr[i];
        }
        arr[i] = pivot;
        sort(arr, low, i - 1);
        sort(arr, i + 1, high);
    }

    public static void main(String[] args) {

    }

}
